subsys/bluetooth/host/classic/sco.c

```diff
@@ -0,0 +1,409 @@
+/* sco.c - Bluetooth sco handling */
+
+/*
+ * Copyright (c) 2015-2016 Intel Corporation
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <errno.h>
+
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/check.h>
+
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+
+#include "common/bt_str.h"
+
+#include "host/addr_internal.h"
+#include "host/hci_core.h"
+#include "br.h"
+#include "host/conn_internal.h"
+#include "sco_internal.h"
+
+#define LOG_LEVEL CONFIG_BT_CONN_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_sco);
+
+struct bt_sco_server *sco_server;
+
+#define SCO_CHAN(_sco) ((_sco)->sco.chan);
+
+int bt_sco_server_register(struct bt_sco_server *server)
+{
+	CHECKIF(!server) {
+		LOG_DBG("Invalid parameter: server %p", server);
+		return -EINVAL;
+	}
+
+	if (sco_server) {
+		return -EADDRINUSE;
+	}
+
+	if (!server->accept) {
+		return -EINVAL;
+	}
+
+#if defined(CONFIG_BT_SMP)
+	if (server->sec_level > BT_SECURITY_L3) {
+		return -EINVAL;
+	}
+#endif /* CONFIG_BT_SMP */
+
+	LOG_DBG("%p", server);
+
+	sco_server = server;
+
+	return 0;
+}
+
+int bt_sco_server_unregister(struct bt_sco_server *server)
+{
+	CHECKIF(!server) {
+		LOG_DBG("Invalid parameter: server %p", server);
+		return -EINVAL;
+	}
+
+	if (sco_server != server) {
+		return -EINVAL;
+	}
+
+	sco_server = NULL;
+
+	return 0;
+}
+
+void bt_sco_connected(struct bt_conn *sco)
+{
+	struct bt_sco_chan *chan;
+
+	if (sco == NULL || sco->type != BT_CONN_TYPE_SCO) {
+		LOG_ERR("Invalid parameters: sco %p sco->type %u", sco, sco ? sco->type : 0);
+		return;
+	}
+
+	LOG_DBG("%p", sco);
+
+	chan = SCO_CHAN(sco);
+
+	if (chan == NULL) {
+		LOG_ERR("Could not lookup chan from connected SCO");
+		return;
+	}
+
+	bt_sco_chan_set_state(chan, BT_SCO_STATE_CONNECTED);
+
+	if (chan->ops && chan->ops->connected) {
+		chan->ops->connected(chan);
+	}
+}
+
+void bt_sco_disconnected(struct bt_conn *sco)
+{
+	struct bt_sco_chan *chan;
+
+	if (sco == NULL || sco->type != BT_CONN_TYPE_SCO) {
+		LOG_ERR("Invalid parameters: sco %p sco->type %u", sco, sco ? sco->type : 0);
+		return;
+	}
+	LOG_DBG("%p", sco);
+
+	chan = SCO_CHAN(sco);
+	if (chan == NULL) {
+		LOG_ERR("Could not lookup chan from connected SCO");
+		return;
+	}
+
+	bt_sco_chan_set_state(chan, BT_SCO_STATE_DISCONNECTED);
+
+	bt_sco_cleanup_acl(sco);
+
+	chan->sco = NULL;
+
+	if (chan->ops && chan->ops->disconnected) {
+		chan->ops->disconnected(chan, sco->err);
+	}
+}
+
+static uint8_t sco_server_check_security(struct bt_conn *conn)
+{
+	if (IS_ENABLED(CONFIG_BT_CONN_DISABLE_SECURITY)) {
+		return BT_HCI_ERR_SUCCESS;
+	}
+
+#if defined(CONFIG_BT_SMP)
+	if (conn->sec_level >= sco_server->sec_level) {
+		return BT_HCI_ERR_SUCCESS;
+	}
+
+	return BT_HCI_ERR_INSUFFICIENT_SECURITY;
+#else
```

Comment:
Same here. BR/EDR security should not depend on SMP in any way (well, except for cross-transport key derivation, but that's a different story)